GEODE-9268: Fix PdxInstance handling after cluster restart

gaussianrecurrence opened the pull request:
- In scenarios where an PdxInstance is obtained and later used, it
   might happen a coredump if the PdxTypeRegistry is cleaned up after the
   PdxInstance is obtained. This happens on those scenarios where
   redundancy is completly lost.
 - This change refactors PdxInstance handling so in all cases, the PdxType
   used is the one owned by the PdxInstance.
   Whenever the PdxInstance is to be written, a check is executed to
   ensure the cluster is aware of the PdxType and if not, register it.
 - Removed PdxInstance serialization retries, as with the new approach
   is not needed anymore.
 - Removed UnknownPdxTypeException exception. Instead whenever a PdxType
   is requested and not present, an IllegalStateException exception is
   thrown, as it happens in the Java client.
 - Now PdxInstance is not serialized whenever created. Instead, it's PDX
   byte stream is generated on-demand. Note that the PdxInstance will be
   serialized before being put into a server, as it was done before this
   change.
 - Fixed TcrMessage deserialization whenever a PdxType is requested by
   its ID and no PdxType was found.
 - Fixed incPdxInstanceCreations so it's incremented strictly whenver a
   PdxInstance is created, and not whenever a PdxInstance is
   deserialized.
 - Fixed IT PdxTypeRegistry cleanupOnClusterRestart logic and renamed to
   cleanupOnClusterRestartAndPut. This test was supposed to verify that
   if a PdxInstance is created, and after that the cluster is restarted,
   there is no coredump while writting it to a region. Instead it was
   creating a PdxInstance before and after the cluster, but it has been
   fixed to work as initially intended.
 - Created a new IT PdxTypeRegistry cleanupOnClusterRestartAndFetchFields
   to verify the issue described in the first bullet is not causing a coredump.
 - Removed old IT testThinClientPdxInstance TS as there is a equivalent new
   IT TS named PdxInstanceTest.
 - Fixed some ITs to work accordingly the the new code.

pdxcodemonkey commented on tests/cpp/testobject/NestedPdxObject.cpp:
```diff
@@ -68,6 +68,31 @@ bool ChildPdx::equals(ChildPdx& other) const {
+      m_childPdx{},
+      m_enum{},
+      m_char{L'\0'},
+      m_charArray{} {}
```
Probably another opportunity to rename member variables to conform to style guide.

pdxcodemonkey reviewed (changes requested):
Looks good overall.  Nothing major to fix, just mostly style stuff and a couple of things to clarify.  Thanks!

gaussianrecurrence commented on cppcache/integration/test/PdxSerializerTest.cpp:
```diff
@@ -115,18 +113,19 @@ std::shared_ptr<Region> setupRegion(Cache& cache) {
+  EXPECT_TRUE(nonPdxType);
 
-  EXPECT_NE(expected, nonPdxType);
   EXPECT_EQ(*expected, *nonPdxType);
```
That's odd, I've used EXPECT_EQ lots of times with std::string. The way EXPECT_EQ internally works is by calling operator== and if the comparison fails, then  both objects are serialized and shown the expected and actual in the output. Problem is whenever there is no serialization possible for a pair of objects you are comparing. In this case you would see an hexdump of the objects that can't be serialized.

Serialization btw is done by calling the ostream& operator<<(ostream& os, const ClassName& dt), so internally gtest uses SFINAE to check whether there is or not an implementation for the specific object being serialized.

gaussianrecurrence commented on cppcache/src/PdxLocalWriter.cpp:
```diff
@@ -330,13 +330,13 @@ PdxWriter& PdxLocalWriter::markIdentityField(const std::string&) {
+  int32_t len = PdxHelper::readInt32(buffer);
+  buffer += 8;
+
+  return {buffer, buffer + len};
```
You've got a pretty good point. I think I'll use the explicit construction, rather than the implicit, so it's more clear what's happening there.

gaussianrecurrence commented on cppcache/integration/test/PdxTypeRegistryTest.cpp:
```diff
@@ -140,19 +141,60 @@ TEST(PdxTypeRegistryTest, cleanupOnClusterRestart) {
   EXPECT_NO_THROW(result = query->execute());
   EXPECT_TRUE(result);
-  EXPECT_GT(result->size(), 0);
+  EXPECT_EQ(result->size(), 1);
```
You've got a pretty good point. I've got so many changes in my head that by mistake I changed this test behaviour. This was working fine as it was. I'll revert this test changes. Also, what do you think about adding some comments to clarify how these tests works? I mean cleanupOnClusterRestart and cleanupOnClusterRestartAndFetchFields

pdxcodemonkey replied:
New comments look great!  It's always helpful to have a little understanding of what the tests are supposed to be doing, esp. for something as complex as PDX.

gaussianrecurrence commented on cppcache/integration/test/PdxTypeRegistryTest.cpp:
```diff
@@ -140,19 +141,60 @@ TEST(PdxTypeRegistryTest, cleanupOnClusterRestart) {
+  }
+
+  listener->waitConnected();
+  auto fields = pdx->getFieldNames();
```
Indeed, you've got a point, that at first glance it'd be hard to understand why this test exists. I think it would be best to add a comment with its description.
Thing is what's being tested here is the following:

- You create a PdxInstance called pdx
- Cluster is restarted and the PdxTypeRegister is cleaned up.
- You try to traverse the PdxInstance and you expect there are no coredumps. In this case getFieldNames is used for example.

Thing is, prior to this change, getFieldNames as many other methods tried to fetch the PdxType by querying the register with its ID, and given it was just cleaned up, it was returning a nullptr, hence coredump.

Being all of this said. You think it would be best to check all of the methods that reads the PdxInstance content?

pdxcodemonkey replied:
For completeness' sake, probably yes?  If we modified a method because it could hit the registry and core dump before, we should probably call it now and make sure it no longer blows up.

gaussianrecurrence replied:
That's already covered within revision 1 :)

gaussianrecurrence commented on cppcache/integration/test/PdxTypeRegistryTest.cpp:
```diff
@@ -140,19 +141,60 @@ TEST(PdxTypeRegistryTest, cleanupOnClusterRestart) {
+    fields_set.insert(field->toString());
+  }
+
+  EXPECT_EQ(fields_set.count("entryName"), 1);
```
This is not the purpose of the test, as it was originally designed

pdxcodemonkey replied:
Fair enough

gaussianrecurrence commented on cppcache/src/PdxHelper.cpp:
```diff
@@ -175,61 +163,59 @@ std::shared_ptr<PdxSerializable> PdxHelper::deserializePdx(DataInput& dataInput,
   } else {
-    // type not found; need to get from server
-    if (!pType) {
+    if (pType == nullptr) {
```
Indeed, sorry I try to leave aside my coding conventions hehe, but sometimes I don't notice. Will change it

gaussianrecurrence commented on cppcache/src/PdxInstanceImpl.hpp:
```diff
@@ -211,35 +210,36 @@ class PdxInstanceImpl : public WritablePdxInstance {
-  std::vector<uint8_t> m_buffer;
-  int m_typeId;
+  mutable std::vector<uint8_t> m_buffer;
+  mutable int32_t m_typeId;
```
This is by convention member variables that are lazily initializated are supposed to be mutable. The reason is because once initialized they are supposed to be immutable, but sadly there is no support from the language for lazy initialization, so that's the way it's usually done.

pdxcodemonkey commented:
@gaussianrecurrence Any chance we can polish this off soon?  It keeps sitting around a couple of weeks at a time, and I think it's close to done.

gaussianrecurrence commented:
> @gaussianrecurrence Any chance we can polish this off soon? It keeps sitting around a couple of weeks at a time, and I think it's close to done.

I am looking into a couple of ITs failing. Once sorted I'll push the changes you suggested to continue with the review process

pdxcodemonkey commented:
@gaussianrecurrence Where are we at with this one?  I'd like to clear this one and #812 ASAP, before they celebrate a birthday :).  They've been sitting around waaaay too long...

gaussianrecurrence commented:
> @gaussianrecurrence Where are we at with this one? I'd like to clear this one and #812 ASAP, before they celebrate a birthday :). They've been sitting around waaaay too long...

Let's try to avoid having to celebrate the PR's birthday :)
Apparently, TransactionCleaningTest.txWithStoppedServer is failing from time to time. So, #923 might need some attention before being able to merge any of the other PRs.

pdxcodemonkey commented:
@gaussianrecurrence Thanks for picking this back up.  I have resolved a few of the comments that you replied to and don't need code changes.  Let's clean up the last few and get this merged!

gaussianrecurrence commented:
> @gaussianrecurrence Thanks for picking this back up. I have resolved a few of the comments that you replied to and don't need code changes. Let's clean up the last few and get this merged!

I've created revision 1 addressing all your comments and improved some things on the testing part. Glad to hear any other feedback you might have :)

pdxcodemonkey commented on cppcache/integration/test/PdxInstanceTest.cpp:
```diff
@@ -273,32 +270,51 @@ TEST(PdxInstanceTest, testPdxInstance) {
 
-  EXPECT_EQ(1, cachePerfStats.getPdxInstanceDeserializations())
+  EXPECT_EQ(0, cachePerfStats.getPdxInstanceDeserializations())
       << "pdxInstanceDeserialization should be equal to 1.";
```
Likewise on line 313, the test is "less than 0", and the text says "greater than 0"
